src/runtime/io.cpp

```diff
@@ -187,23 +194,45 @@ extern "C" obj_res lean_io_prim_handle_write(b_obj_arg h, obj_arg buf, obj_arg /
     if (m == n) {
         return set_io_result(box(0));
     } else {
-        return set_io_error(box(lean_errno_to_io_error(errno)));
+        return set_io_error(lean_errno_to_io_error(errno));
     }
 }
 
-/* Handle.get_line : Handle → IO Unit */
-extern "C" obj_res lean_io_prim_handle_get_line(b_obj_arg /* h */, obj_arg /* w */) {
-    // not implemented yet
-    lean_unreachable();
+/* Handle.getLine : Handle → IO Unit                     */
+/*   The line returned by `lean_io_prim_handle_get_line` */
+/*   is truncated at the first '\0' character and the    */
+/*   rest of the line is discarded.                      */
+extern "C" obj_res lean_io_prim_handle_get_line(b_obj_arg h, obj_arg /* w */) {
+    FILE * fp = io_get_handle(h);
+    usize n = 0;          // will receive the size of buf
+    char * buf = nullptr; // will receive the buffer allocated by getline
+    if (feof(fp)) {
+        return set_io_error(lean_io_error_eof);
+    }
+    int r = getline(&buf, &n, fp);
+    if (r != -1) {
+        object * res = lean_mk_string(buf);
+        free(buf);
+        return set_io_result(res);
+    } else if (feof(fp)) {
+        free(buf);
+        return set_io_result(lean_mk_string(""));
+    } else {
+        return set_io_error(lean_errno_to_io_error(errno));
+    }
 }
 
-/* Handle.get_line : Handle → IO Unit */
-extern "C" obj_res lean_io_prim_handle_put_str(b_obj_arg, obj_arg /* w */) {
-    // not implemented yet
-    lean_unreachable();
+/* Handle.putStr : Handle → String → IO Unit */
+extern "C" obj_res lean_io_prim_handle_put_str(b_obj_arg h, obj_arg s, obj_arg /* w */) {
+    FILE * fp = io_get_handle(h);
+    if (std::fputs(lean_string_cstr(s), fp)) {
+        return set_io_result(box(0));
+    } else {
+        return set_io_error(lean_errno_to_io_error(errno));
```

Comment:
We should use `ferror()` instead of `errno` https://wiki.sei.cmu.edu/confluence/display/c/ERR01-C.+Use+ferror%28%29+rather+than+errno+to+check+for+FILE+stream+errors

Reply:
I think the point of that article is that `errno == 0` / `errno != 0` is not a good way of checking whether an error occurs or not. My code uses the return value of the functions I'm wrapping to decide whether an error occurred. Then, I use `errno` to determine which error occurred and return a value of an inductive type to encode the information. This is not something `ferror` can do.

Reply:
Oh wow, I missed that `ferror` doesn't return the same codes as `errno`. Ok.